File: python-interpreter/untildeath/builtins.py
"""Built-in rites (functions) for !~ATH."""

import random
import sys
import time
from typing import Any, List

from .errors import RuntimeError


def stringify(value: Any) -> str:
    """Convert a value to its string representation."""
    if value is None:
        return "VOID"
    if isinstance(value, bool):
        return "ALIVE" if value else "DEAD"
    if isinstance(value, list):
        return "[" + ", ".join(stringify(v) for v in value) + "]"
    if isinstance(value, dict):
        entries = ", ".join(f"{k}: {stringify(v)}" for k, v in value.items())
        return "{" + entries + "}"
    return str(value)


def type_name(value: Any) -> str:
    """Get the type name of a value."""
    if value is None:
        return "VOID"
    if isinstance(value, bool):
        return "BOOLEAN"
    if isinstance(value, int):
        return "INTEGER"
    if isinstance(value, float):
        return "FLOAT"
    if isinstance(value, str):
        return "STRING"
    if isinstance(value, list):
        return "ARRAY"
    if isinstance(value, dict):
        return "MAP"
    return "UNKNOWN"


def is_truthy(value: Any) -> bool:
    """Determine if a value is truthy."""
    if value is None:
        return False
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return value != 0
    if isinstance(value, str):
        return len(value) > 0
    if isinstance(value, (list, dict)):
        return len(value) > 0
    return True


class Builtins:
    """Container for all built-in rites."""

    def __init__(self, interpreter):
        self.interpreter = interpreter
        self._input_buffer = []

    def get(self, name: str):
        """Get a built-in function by name."""
        builtins = {
            # I/O
            'UTTER': self.utter,
            'HEED': self.heed,
            'SCRY': self.scry,
            'INSCRIBE': self.inscribe,

            # Type operations
            'TYPEOF': self.typeof,
            'LENGTH': self.length,
            'PARSE_INT': self.parse_int,
            'PARSE_FLOAT': self.parse_float,
            'STRING': self.string,
            'INT': self.int_,
            'FLOAT': self.float_,
            'CHAR': self.char,
            'CODE': self.code,
            'BIN': self.bin,
            'HEX': self.hex,

            # Array operations
            'APPEND': self.append,
            'PREPEND': self.prepend,
            'SLICE': self.slice,
            'FIRST': self.first,
            'LAST': self.last,
            'CONCAT': self.concat,

            # Map operations
            'KEYS': self.keys,
            'VALUES': self.values,
            'HAS': self.has,
            'SET': self.set_,
            'DELETE': self.delete,

            # String operations
            'SPLIT': self.split,
            'JOIN': self.join,
            'SUBSTRING': self.substring,
            'UPPERCASE': self.uppercase,
            'LOWERCASE': self.lowercase,
            'TRIM': self.trim,
            'REPLACE': self.replace,

            # Utility
            'RANDOM': self.random,
            'RANDOM_INT': self.random_int,
            'TIME': self.time,
        }
        return builtins.get(name)

    # ============ I/O ============

    def utter(self, *args) -> None:
        """Print values to stdout."""
        output = " ".join(stringify(arg) for arg in args)
        print(output)
        return None

    def heed(self) -> str:
        """Read a line from stdin."""
        try:
            return input()
        except EOFError:
            return ""

    def scry(self, path: Any) -> str:
        """Read file contents or stdin."""
        if path is None:
            return sys.stdin.read()

        if not isinstance(path, str):
            raise RuntimeError(f"SCRY expects string path or VOID, got {type_name(path)}")
        try:
            with open(path, 'r', encoding='utf-8') as f:
                return f.read()
        except FileNotFoundError:
            raise RuntimeError(f"File not found: {path}")
        except IOError as e:
            raise RuntimeError(f"Cannot read file: {e}")

    def inscribe(self, path: str, content: str) -> None:
        """Write content to file."""
        if not isinstance(path, str):
            raise RuntimeError(f"INSCRIBE expects string path, got {type_name(path)}")
        if not isinstance(content, str):
            content = stringify(content)
        try:
            with open(path, 'w', encoding='utf-8') as f:
                f.write(content)
        except IOError as e:
            raise RuntimeError(f"Cannot write file: {e}")
        return None

    # ============ Type Operations ============

    def typeof(self, value: Any) -> str:
        """Get type name of value."""
        return type_name(value)

    def length(self, value: Any) -> int:
        """Get length of string or array."""
        if isinstance(value, (str, list)):
            return len(value)
        raise RuntimeError(f"LENGTH expects string or array, got {type_name(value)}")

    def parse_int(self, value: str) -> int:
        """Parse string to integer."""
        if not isinstance(value, str):
            raise RuntimeError(f"PARSE_INT expects string, got {type_name(value)}")
        try:
            # Ensure it's a valid integer (not float)
            if '.' in value:
                raise ValueError("Float string")
            return int(value)
        except ValueError:
            raise RuntimeError(f"Cannot parse '{value}' as integer")

    def parse_float(self, value: str) -> float:
        """Parse string to float."""
        if not isinstance(value, str):
            raise RuntimeError(f"PARSE_FLOAT expects string, got {type_name(value)}")
        try:
            return float(value)
        except ValueError:
            raise RuntimeError(f"Cannot parse '{value}' as float")

    def string(self, value: Any) -> str:
        """Convert value to string."""
        return stringify(value)

    def int_(self, value: Any) -> int:
        """Convert float to integer (truncate)."""
        if isinstance(value, float):
            return int(value)
        if isinstance(value, int):
            return value
        raise RuntimeError(f"INT expects number, got {type_name(value)}")

    def float_(self, value: Any) -> float:
        """Convert integer to float."""
        if isinstance(value, (int, float)):
            return float(value)
        raise RuntimeError(f"FLOAT expects number, got {type_name(value)}")

    def char(self, value: int) -> str:
        """Convert integer code point to character."""
        if not isinstance(value, int):
            raise RuntimeError(f"CHAR expects integer, got {type_name(value)}")
        try:
            return chr(value)
        except ValueError:
            raise RuntimeError(f"Invalid code point: {value}")

    def code(self, value: str) -> int:
        """Get integer code point of first character."""
        if not isinstance(value, str):
            raise RuntimeError(f"CODE expects string, got {type_name(value)}")
        if len(value) == 0:
            raise RuntimeError("CODE called on empty string")
        return ord(value[0])

    def bin(self, value: int) -> str:
        """Convert integer to binary string."""
        if not isinstance(value, int):
            raise RuntimeError(f"BIN expects integer, got {type_name(value)}")
        return bin(value)[2:]  # Strip '0b' prefix

    def hex(self, value: int) -> str:
        """Convert integer to hex string."""
        if not isinstance(value, int):
            raise RuntimeError(f"HEX expects integer, got {type_name(value)}")
        return hex(value)[2:].upper()  # Strip '0x' prefix and uppercase

    # ============ Array Operations ============

    def append(self, arr: list, value: Any) -> list:
        """Add element to end of array."""
        if not isinstance(arr, list):
            raise RuntimeError(f"APPEND expects array, got {type_name(arr)}")
        return arr + [value]

    def prepend(self, arr: list, value: Any) -> list:
        """Add element to beginning of array."""
        if not isinstance(arr, list):
            raise RuntimeError(f"PREPEND expects array, got {type_name(arr)}")
        return [value] + arr

    def slice(self, arr: list, start: int, end: int) -> list:
        """Extract subsequence from array."""
        if not isinstance(arr, list):
            raise RuntimeError(f"SLICE expects array, got {type_name(arr)}")
        if not isinstance(start, int) or not isinstance(end, int):
            raise RuntimeError("SLICE expects integer indices")
        return arr[start:end]

    def first(self, arr: list) -> Any:
        """Get first element of array."""
        if not isinstance(arr, list):
            raise RuntimeError(f"FIRST expects array, got {type_name(arr)}")
        if len(arr) == 0:
            raise RuntimeError("FIRST called on empty array")
        return arr[0]

    def last(self, arr: list) -> Any:
        """Get last element of array."""
        if not isinstance(arr, list):
            raise RuntimeError(f"LAST expects array, got {type_name(arr)}")
        if len(arr) == 0:
            raise RuntimeError("LAST called on empty array")
        return arr[-1]

    def concat(self, arr1: list, arr2: list) -> list:
        """Concatenate two arrays."""
        if not isinstance(arr1, list) or not isinstance(arr2, list):
            raise RuntimeError("CONCAT expects two arrays")
        return arr1 + arr2

    # ============ Map Operations ============

    def keys(self, m: dict) -> list:
        """Get array of map keys."""
        if not isinstance(m, dict):
            raise RuntimeError(f"KEYS expects map, got {type_name(m)}")
        return list(m.keys())

    def values(self, m: dict) -> list:
        """Get array of map values."""
        if not isinstance(m, dict):
            raise RuntimeError(f"VALUES expects map, got {type_name(m)}")
        return list(m.values())

    def has(self, m: dict, key: str) -> bool:
        """Check if map has key."""
        if not isinstance(m, dict):
            raise RuntimeError(f"HAS expects map, got {type_name(m)}")
        return key in m

    def set_(self, m: dict, key: str, value: Any) -> dict:
        """Set key-value pair in map."""
        if not isinstance(m, dict):
            raise RuntimeError(f"SET expects map, got {type_name(m)}")
        result = m.copy()
        result[key] = value
        return result

    def delete(self, m: dict, key: str) -> dict:
        """Remove key from map."""
        if not isinstance(m, dict):
            raise RuntimeError(f"DELETE expects map, got {type_name(m)}")
        result = m.copy()
        result.pop(key, None)
        return result

    # ============ String Operations ============

    def split(self, s: str, delimiter: str) -> list:
        """Split string into array."""
        if not isinstance(s, str) or not isinstance(delimiter, str):
            raise RuntimeError("SPLIT expects two strings")
        if delimiter == "":
            return list(s)
        return s.split(delimiter)

    def join(self, arr: list, delimiter: str) -> str:
        """Join array into string."""
        if not isinstance(arr, list):
            raise RuntimeError(f"JOIN expects array, got {type_name(arr)}")
        if not isinstance(delimiter, str):
            raise RuntimeError(f"JOIN expects string delimiter, got {type_name(delimiter)}")
        return delimiter.join(stringify(v) if not isinstance(v, str) else v for v in arr)

    def substring(self, s: str, start: int, end: int) -> str:
        """Extract substring."""
        if not isinstance(s, str):
            raise RuntimeError(f"SUBSTRING expects string, got {type_name(s)}")
        if not isinstance(start, int) or not isinstance(end, int):
            raise RuntimeError("SUBSTRING expects integer indices")
        return s[start:end]

    def uppercase(self, s: str) -> str:
        """Convert to uppercase."""
        if not isinstance(s, str):
            raise RuntimeError(f"UPPERCASE expects string, got {type_name(s)}")
        return s.upper()

    def lowercase(self, s: str) -> str:
        """Convert to lowercase."""
        if not isinstance(s, str):
            raise RuntimeError(f"LOWERCASE expects string, got {type_name(s)}")
        return s.lower()

    def trim(self, s: str) -> str:
        """Remove leading/trailing whitespace."""
        if not isinstance(s, str):
            raise RuntimeError(f"TRIM expects string, got {type_name(s)}")
        return s.strip()

    def replace(self, s: str, old: str, new: str) -> str:
        """Replace occurrences in string."""
        if not isinstance(s, str) or not isinstance(old, str) or not isinstance(new, str):
            raise RuntimeError("REPLACE expects three strings")
        return s.replace(old, new)

    # ============ Utility ============

    def random(self) -> float:
        """Random float between 0 and 1."""
        return random.random()

    def random_int(self, min_val: int, max_val: int) -> int:
        """Random integer in range (inclusive)."""
        if not isinstance(min_val, int) or not isinstance(max_val, int):
            raise RuntimeError("RANDOM_INT expects two integers")
        return random.randint(min_val, max_val)

    def time(self) -> int:
        """Current Unix timestamp in milliseconds."""
        return int(time.time() * 1000)

File: js-interpreter/src/builtins.js
/**
 * Built-in rites (functions) for !~ATH.
 */

import { RuntimeError } from './errors.js';

/**
 * Convert a value to its string representation.
 */
export function stringify(value) {
  if (value === null || value === undefined) {
    return 'VOID';
  }
  if (typeof value === 'boolean') {
    return value ? 'ALIVE' : 'DEAD';
  }
  if (Array.isArray(value)) {
    return '[' + value.map(stringify).join(', ') + ']';
  }
  if (typeof value === 'object') {
    const entries = Object.entries(value)
      .map(([k, v]) => `${k}: ${stringify(v)}`)
      .join(', ');
    return '{' + entries + '}';
  }
  return String(value);
}

/**
 * Get the type name of a value.
 */
export function typeName(value) {
  if (value === null || value === undefined) {
    return 'VOID';
  }
  if (typeof value === 'boolean') {
    return 'BOOLEAN';
  }
  if (typeof value === 'number') {
    return Number.isInteger(value) ? 'INTEGER' : 'FLOAT';
  }
  if (typeof value === 'string') {
    return 'STRING';
  }
  if (Array.isArray(value)) {
    return 'ARRAY';
  }
  if (typeof value === 'object') {
    return 'MAP';
  }
  return 'UNKNOWN';
}

/**
 * Determine if a value is truthy.
 */
export function isTruthy(value) {
  if (value === null || value === undefined) {
    return false;
  }
  if (typeof value === 'boolean') {
    return value;
  }
  if (typeof value === 'number') {
    return value !== 0;
  }
  if (typeof value === 'string') {
    return value.length > 0;
  }
  if (Array.isArray(value)) {
    return value.length > 0;
  }
  if (typeof value === 'object') {
    return Object.keys(value).length > 0;
  }
  return true;
}

/**
 * Container for all built-in rites.
 */
export class Builtins {
  constructor(interpreter, options = {}) {
    this.interpreter = interpreter;
    this.onOutput = options.onOutput || ((text) => console.log(text));
    this.onInput = options.onInput || null;
    this.inputQueue = options.inputQueue || [];
    this._inputIndex = 0;
    this.scryQueue = options.scryQueue || [];
    this._scryIndex = 0;
  }

  get(name) {
    const builtins = {
      // I/O
      'UTTER': (...args) => this.utter(...args),
      'HEED': () => this.heed(),
      'SCRY': (path) => this.scry(path),

      // Type operations
      'TYPEOF': (value) => this.typeof_(value),
      'LENGTH': (value) => this.length(value),
      'PARSE_INT': (value) => this.parseIntFunc(value),
      'PARSE_FLOAT': (value) => this.parseFloatFunc(value),
      'STRING': (value) => this.string(value),
      'INT': (value) => this.int(value),
      'FLOAT': (value) => this.float(value),
      'CHAR': (value) => this.char(value),
      'CODE': (value) => this.code(value),
      'BIN': (value) => this.bin(value),
      'HEX': (value) => this.hex(value),

      // Array operations
      'APPEND': (arr, value) => this.append(arr, value),
      'PREPEND': (arr, value) => this.prepend(arr, value),
      'SLICE': (arr, start, end) => this.slice(arr, start, end),
      'FIRST': (arr) => this.first(arr),
      'LAST': (arr) => this.last(arr),
      'CONCAT': (arr1, arr2) => this.concat(arr1, arr2),

      // Map operations
      'KEYS': (m) => this.keys(m),
      'VALUES': (m) => this.values(m),
      'HAS': (m, key) => this.has(m, key),
      'SET': (m, key, value) => this.set(m, key, value),
      'DELETE': (m, key) => this.delete(m, key),

      // String operations
      'SPLIT': (s, delimiter) => this.split(s, delimiter),
      'JOIN': (arr, delimiter) => this.join(arr, delimiter),
      'SUBSTRING': (s, start, end) => this.substring(s, start, end),
      'UPPERCASE': (s) => this.uppercase(s),
      'LOWERCASE': (s) => this.lowercase(s),
      'TRIM': (s) => this.trim(s),
      'REPLACE': (s, old, newStr) => this.replace(s, old, newStr),

      // Utility
      'RANDOM': () => this.random(),
      'RANDOM_INT': (min, max) => this.randomInt(min, max),
      'TIME': () => this.time(),
    };
    return builtins[name] || null;
  }

  // ============ I/O ============

  utter(...args) {
    const output = args.map(stringify).join(' ');
    this.onOutput(output);
    return null;
  }

  heed() {
    if (this.inputQueue.length > this._inputIndex) {
      return this.inputQueue[this._inputIndex++];
    }
    if (this.onInput) {
      return this.onInput();
    }
    return '';
  }

  scry(path) {
    if (path !== null && path !== undefined) {
      throw new RuntimeError(`SCRY expects VOID (for stdin), got ${typeName(path)}`);
    }
    if (this.scryQueue.length > this._scryIndex) {
      return this.scryQueue[this._scryIndex++];
    }
    return '';
  }

  // ============ Type Operations ============

  typeof_(value) {
    return typeName(value);
  }

  length(value) {
    if (typeof value === 'string' || Array.isArray(value)) {
      return value.length;
    }
    throw new RuntimeError(`LENGTH expects string or array, got ${typeName(value)}`);
  }

  parseIntFunc(value) {
    if (typeof value !== 'string') {
      throw new RuntimeError(`PARSE_INT expects string, got ${typeName(value)}`);
    }
    if (value.includes('.')) {
      throw new RuntimeError(`Cannot parse '${value}' as integer`);
    }
    const result = parseInt(value, 10);
    if (isNaN(result)) {
      throw new RuntimeError(`Cannot parse '${value}' as integer`);
    }
    return result;
  }

  parseFloatFunc(value) {
    if (typeof value !== 'string') {
      throw new RuntimeError(`PARSE_FLOAT expects string, got ${typeName(value)}`);
    }
    const result = parseFloat(value);
    if (isNaN(result)) {
      throw new RuntimeError(`Cannot parse '${value}' as float`);
    }
    return result;
  }

  string(value) {
    return stringify(value);
  }

  int(value) {
    if (typeof value === 'number') {
      return Math.trunc(value);
    }
    throw new RuntimeError(`INT expects number, got ${typeName(value)}`);
  }

  float(value) {
    if (typeof value === 'number') {
      return value;
    }
    throw new RuntimeError(`FLOAT expects number, got ${typeName(value)}`);
  }

  char(value) {
    if (!Number.isInteger(value)) {
      throw new RuntimeError(`CHAR expects integer, got ${typeName(value)}`);
    }
    try {
      return String.fromCodePoint(value);
    } catch (e) {
      throw new RuntimeError(`Invalid code point: ${value}`);
    }
  }

  code(value) {
    if (typeof value !== 'string') {
      throw new RuntimeError(`CODE expects string, got ${typeName(value)}`);
    }
    if (value.length === 0) {
      throw new RuntimeError('CODE called on empty string');
    }
    return value.codePointAt(0);
  }

  bin(value) {
    if (!Number.isInteger(value)) {
      throw new RuntimeError(`BIN expects integer, got ${typeName(value)}`);
    }
    return (value >>> 0).toString(2); // Unsigned binary string representation
  }

  hex(value) {
    if (!Number.isInteger(value)) {
      throw new RuntimeError(`HEX expects integer, got ${typeName(value)}`);
    }
    return (value >>> 0).toString(16).toUpperCase(); // Unsigned hex string
  }

  // ============ Array Operations ============

  append(arr, value) {
    if (!Array.isArray(arr)) {
      throw new RuntimeError(`APPEND expects array, got ${typeName(arr)}`);
    }
    return [...arr, value];
  }

  prepend(arr, value) {
    if (!Array.isArray(arr)) {
      throw new RuntimeError(`PREPEND expects array, got ${typeName(arr)}`);
    }
    return [value, ...arr];
  }

  slice(arr, start, end) {
    if (!Array.isArray(arr)) {
      throw new RuntimeError(`SLICE expects array, got ${typeName(arr)}`);
    }
    if (typeof start !== 'number' || typeof end !== 'number') {
      throw new RuntimeError('SLICE expects integer indices');
    }
    return arr.slice(start, end);
  }

  first(arr) {
    if (!Array.isArray(arr)) {
      throw new RuntimeError(`FIRST expects array, got ${typeName(arr)}`);
    }
    if (arr.length === 0) {
      throw new RuntimeError('FIRST called on empty array');
    }
    return arr[0];
  }

  last(arr) {
    if (!Array.isArray(arr)) {
      throw new RuntimeError(`LAST expects array, got ${typeName(arr)}`);
    }
    if (arr.length === 0) {
      throw new RuntimeError('LAST called on empty array');
    }
    return arr[arr.length - 1];
  }

  concat(arr1, arr2) {
    if (!Array.isArray(arr1) || !Array.isArray(arr2)) {
      throw new RuntimeError('CONCAT expects two arrays');
    }
    return [...arr1, ...arr2];
  }

  // ============ Map Operations ============

  keys(m) {
    if (typeof m !== 'object' || m === null || Array.isArray(m)) {
      throw new RuntimeError(`KEYS expects map, got ${typeName(m)}`);
    }
    return Object.keys(m);
  }

  values(m) {
    if (typeof m !== 'object' || m === null || Array.isArray(m)) {
      throw new RuntimeError(`VALUES expects map, got ${typeName(m)}`);
    }
    return Object.values(m);
  }

  has(m, key) {
    if (typeof m !== 'object' || m === null || Array.isArray(m)) {
      throw new RuntimeError(`HAS expects map, got ${typeName(m)}`);
    }
    return key in m;
  }

  set(m, key, value) {
    if (typeof m !== 'object' || m === null || Array.isArray(m)) {
      throw new RuntimeError(`SET expects map, got ${typeName(m)}`);
    }
    return { ...m, [key]: value };
  }

  delete(m, key) {
    if (typeof m !== 'object' || m === null || Array.isArray(m)) {
      throw new RuntimeError(`DELETE expects map, got ${typeName(m)}`);
    }
    const result = { ...m };
    delete result[key];
    return result;
  }

  // ============ String Operations ============

  split(s, delimiter) {
    if (typeof s !== 'string' || typeof delimiter !== 'string') {
      throw new RuntimeError('SPLIT expects two strings');
    }
    if (delimiter === '') {
      return [...s];
    }
    return s.split(delimiter);
  }

  join(arr, delimiter) {
    if (!Array.isArray(arr)) {
      throw new RuntimeError(`JOIN expects array, got ${typeName(arr)}`);
    }
    if (typeof delimiter !== 'string') {
      throw new RuntimeError(`JOIN expects string delimiter, got ${typeName(delimiter)}`);
    }
    return arr.map(v => typeof v === 'string' ? v : stringify(v)).join(delimiter);
  }

  substring(s, start, end) {
    if (typeof s !== 'string') {
      throw new RuntimeError(`SUBSTRING expects string, got ${typeName(s)}`);
    }
    if (typeof start !== 'number' || typeof end !== 'number') {
      throw new RuntimeError('SUBSTRING expects integer indices');
    }
    return s.slice(start, end);
  }

  uppercase(s) {
    if (typeof s !== 'string') {
      throw new RuntimeError(`UPPERCASE expects string, got ${typeName(s)}`);
    }
    return s.toUpperCase();
  }

  lowercase(s) {
    if (typeof s !== 'string') {
      throw new RuntimeError(`LOWERCASE expects string, got ${typeName(s)}`);
    }
    return s.toLowerCase();
  }

  trim(s) {
    if (typeof s !== 'string') {
      throw new RuntimeError(`TRIM expects string, got ${typeName(s)}`);
    }
    return s.trim();
  }

  replace(s, old, newStr) {
    if (typeof s !== 'string' || typeof old !== 'string' || typeof newStr !== 'string') {
      throw new RuntimeError('REPLACE expects three strings');
    }
    return s.split(old).join(newStr);
  }

  // ============ Utility ============

  random() {
    return Math.random();
  }

  randomInt(min, max) {
    if (typeof min !== 'number' || typeof max !== 'number') {
      throw new RuntimeError('RANDOM_INT expects two integers');
    }
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  time() {
    return Date.now();
  }
}
